test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+    metadata.setOutputArray(af::reorder(metadata.out_arr,
+                                        reorder_0, reorder_1, reorder_2));
+}
+
+// Partner function of testWriteToOutputArray. This generates the "special"
+// array that testWriteToOutputArray will use to check if the af_* function
+// correctly uses an existing array as its output
+void genTestOutputArray(af_array *out_ptr,
+                        const unsigned ndims, const dim_t *dims,
+                        const af::dtype ty,
+                        TestOutputArrayInfo* metadata) {
+    af::dim4 arr_dims(ndims, dims);
+    switch (metadata->getOutputArrayType()) {
+    case NULL_ARRAY:
+        genNullArray(*metadata, arr_dims, ty);
+        break;
+    case FULL_ARRAY:
+        genRegularArray(*metadata, arr_dims, ty);
+        break;
+    case SUB_ARRAY:
+        genSubArray(*metadata, arr_dims, ty);
+        break;
+    case REORDERED_ARRAY:
+        genReorderedArray(*metadata, arr_dims, ty);
+        break;
+    }
+    *out_ptr = metadata->getOutputArrayC();
+}
+
+// Partner function of genTestOutputArray. This uses the same "special"
+// array that genTestOutputArray generates, and checks whether the
+// af_* function wrote to that array correctly
+::testing::AssertionResult
+testWriteToOutputArray(std::string gold_name, std::string result_name,
+                       const af_array gold, const af_array out,
+                       TestOutputArrayInfo *metadata) {
+    // In the case of NULL_ARRAY, the output array starts out as null.
+    // After the af_* function is called, it shouldn't be null anymore
+    if (metadata->getOutputArrayType() == NULL_ARRAY) {
+        if (out == 0) {
+            return ::testing::AssertionFailure()
+                << "Output af_array " << result_name << " is null";
+        }
+        metadata->setOutputArrayC(out);
+    }
+    // For every other case, must check if the af_array generated by
+    // genTestOutputArray was used by the af_* function as its output array
+    else {
+        if (metadata->getOutputArrayC() != out) {
+            return ::testing::AssertionFailure()
+                << "af_array POINTER MISMATCH:\n"
+                << "  Actual: " << out << "\n"
+                << "Expected: " << metadata->getOutputArrayC();
+        }
+    }
+
+    if (metadata->getOutputArrayType() == SUB_ARRAY) {
+        // There are two full arrays. One will be injected with the gold
+        // subarray, the other should have already been injected with the af_*
+        // function's output. Then we compare the two full arrays
+
+        af_array gold_retain = 0;
+        af_retain_array(&gold_retain, gold);
+        af::array gold_retain_cpp(gold_retain);
+
+        af::copy(metadata->out_arr_cpy, gold_retain_cpp,
+                 metadata->getSubArrayIdx(0),
+                 metadata->getSubArrayIdx(1),
+                 metadata->getSubArrayIdx(2),
+                 metadata->getSubArrayIdx(3));
```

Comment:
this if block will change to something like the following
```C++
af_array gold_retain = 0;
af_retain_array(&gold_retain, gold);
af::array gold_retain_cpp(gold_retain);
af::array original = metadata->getOutputArray();
af::array subout = original(metadata->idx(0), metadata->idx(1),
                            metadata->idx(2), metadata->idx(3));
af::array subgold = gold_retain_cpp(metadata->idx(0), metadata->idx(1),
                                    metadata->idx(2), metadata->idx(3));
return assertArrayEq(gold_name, result_name, subgold, subarr);
```

Reply:
I think you've got this already from our conversation in Slack, but basically my concern here is reflected in my response to your comment in `genSubArray`. Anyway, `gold` and `gold_retain_cpp` are both the reference sub-array. We want to inject that into a duplicate of the original random array that we generated in `genTestOutputArray, and compare the duplicate to the original (both full arrays) - both must equal element wise, because at this point, `af_approx1` should have already injected its output into the full original random array that we generated.